File: sdk/py/hop_top_kit/output/formatters/json_formatter.py
"""json built-in formatter — stdlib json with configurable indent.

Key order mirrors the tabular formatters: ``columns`` sets it, ``cols``
reorders + selects, payload insertion order is the fallback.
"""

from __future__ import annotations

import dataclasses
import json
from typing import TYPE_CHECKING, Any, TextIO

from hop_top_kit.output.formatter import OptionSpec
from hop_top_kit.output.projection import project_payload

if TYPE_CHECKING:
    from hop_top_kit.output.formatter import ColumnSpec


class JSONFormatter:
    key = "json"
    extensions: tuple[str, ...] = (".json",)

    def options(self) -> list[OptionSpec]:
        return [
            OptionSpec(
                name="indent",
                type="int",
                default=2,
                usage="indent width in spaces (0 disables pretty-print)",
            )
        ]

    def render(
        self,
        out: TextIO,
        data: Any,
        opts: dict[str, Any],
        cols: list[str],
        columns: list[ColumnSpec] | None = None,
    ) -> None:
        indent = opts.get("indent", 2)
        if indent == 0:
            indent = None  # json.dumps None → compact
        keys = cols or [c.header for c in columns or []]
        out.write(json.dumps(_to_jsonable(project_payload(data, keys)), indent=indent))
        out.write("\n")


def _to_jsonable(v: Any) -> Any:
    """Convert dataclass instances to dicts so json.dumps doesn't crash."""
    if dataclasses.is_dataclass(v) and not isinstance(v, type):
        return dataclasses.asdict(v)
    if isinstance(v, list):
        return [_to_jsonable(item) for item in v]
    if isinstance(v, dict):
        return {k: _to_jsonable(val) for k, val in v.items()}
    return v

File: sdk/experimental/php/src/Output/Formatter/Builtin/JsonFormatter.php
<?php

declare(strict_types=1);

namespace HopTop\Kit\Output\Formatter\Builtin;

use HopTop\Kit\Output\Formatter\Formatter;
use HopTop\Kit\Output\Formatter\OptionSpec;
use HopTop\Kit\Output\Formatter\OptionType;
use HopTop\Kit\Output\Formatter\Projection;
use RuntimeException;

/**
 * JSON formatter. Mirrors py/ts/go json built-ins.
 *
 * Options:
 *   - indent (int, default 2) — number of spaces per indent level; 0 = compact
 *
 * Single-row payloads emit a JSON object; list payloads emit a JSON array.
 * Key order follows the resolved $cols (--cols, else the caller's ColumnSpec
 * order), else the payload's own key order — PHP arrays are insertion-
 * ordered, so the resolved order survives json_encode unchanged.
 */
final class JsonFormatter implements Formatter
{
    public function key(): string
    {
        return 'json';
    }

    public function extensions(): array
    {
        return ['.json'];
    }

    public function options(): array
    {
        return [
            new OptionSpec(
                name: 'indent',
                type: OptionType::Int,
                usage: 'Indent width in spaces (0 = compact)',
                default: 2,
            ),
        ];
    }

    /**
     * @param list<string> $cols resolved column projection
     */
    public function render(mixed $writer, mixed $data, array $opts, array $cols): void
    {
        $indent = is_int($opts['indent'] ?? null) ? (int) $opts['indent'] : 2;
        $projected = Projection::project($data, $cols);

        $flags = JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE | JSON_THROW_ON_ERROR;
        if ($indent > 0) {
            $flags |= JSON_PRETTY_PRINT;
        }

        $json = json_encode($projected, $flags);

        // JSON_PRETTY_PRINT hard-codes 4-space indent; rewrite to honor opts.
        if ($indent > 0 && $indent !== 4) {
            $pad = str_repeat(' ', $indent);
            $json = preg_replace_callback(
                '/^( {4})+/m',
                static fn (array $m) => str_repeat($pad, (int) (strlen($m[0]) / 4)),
                $json,
            ) ?? $json;
        }

        if (fwrite($writer, $json . "\n") === false) {
            throw new RuntimeException('json: write failed');
        }
    }

}
